src/simplify_reshapes.cpp

```diff
@@ -900,26 +900,80 @@ struct find_scalar_multibroadcast_reshape_or_transpose
     }
 };
 
-struct find_reshape_reshape_dot
+// Remove unnecessary preceeding size 1 dims for constants
+struct find_const_multibroadcast
 {
     auto matcher() const
     {
-        return match::name("dot")(match::used_once(),
-                                  match::args(match::name("reshape").bind("inp_rsp1"),
-                                              match::name("reshape").bind("inp_rsp2")));
+        return match::name("multibroadcast")(match::arg(0)(match::is_constant().bind("constant")));
+    }
+
+    void apply(module& m, const match::matcher_result& mr) const
+    {
+        auto mbr      = mr.result;
+        auto constant = mr.instructions["constant"];
+
+        if(constant->get_shape().lens().size() <= 1)
+            return;
+
+        auto const_lens = constant->get_shape().lens();
+        auto it = std::find_if(const_lens.begin(), const_lens.end(), [](auto i) { return i != 1; });
+        auto naxes = std::distance(const_lens.begin(), it);
+        if(naxes == 0)
+            return;
+
+        std::vector<std::size_t> sq_axes(naxes);
+        std::iota(sq_axes.begin(), sq_axes.end(), 0);
+
+        auto sq_const =
+            m.insert_instruction(mbr, make_op("squeeze", {{"axes", sq_axes}}), constant);
+        m.replace_instruction(mbr, mbr->get_operator(), sq_const);
+    }
+};
+
+// Move convert before reshape when preceeding a dot op
+struct find_reshape_convert_dot
+{
+    auto matcher() const
+    {
+        auto dot_output = match::any_of[match::outputs()](match::name("dot"));
+        return match::name("convert")(match::arg(0)(match::name("reshape").bind("rsp")),
+                                      dot_output);
+    }
+
+    void apply(module& m, const match::matcher_result& r) const
+    {
+        auto convert = r.result;
+        auto rsp     = r.instructions["rsp"];
+
+        auto inp         = rsp->inputs().front();
+        auto new_convert = m.insert_instruction(rsp, convert->get_operator(), inp);
+        m.replace_instruction(convert, rsp->get_operator(), new_convert);
+    }
+};
```

Comment:
First, we should look for any unary operator not just convert. Secondly we should move past all reshape/transpose/broadcast operators. Thirdly, we may want to move it to the input or the output depending on what is on the input or output. There are 3 cases I see:

* `pointwise/reduce/constant -> reshapes -> unary -> reshapes -> non-pointwise/reduce`: we should move the unary operator to the left(ie the inputs).
* `non-pointwise/reduce -> reshapes -> unary -> reshapes -> pointwise/reduce`: we should move the unary operator to the right(ie the output).
* `pointwise/reduce/constant -> reshapes -> unary -> reshapes -> pointwise/reduce`: we could move the unary to either side but I guess we could prefer the input which might be better if its an input.

Reply:
This is implemented in #2958.

Reply:
So I think this can be removed.

Reply:
Removed